Source/WebCore/Modules/url-pattern/URLPatternCanonical.cpp

```diff
@@ -0,0 +1,258 @@
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "URLPatternCanonical.h"
+
+#include <wtf/text/MakeString.h>
+#include <wtf/text/StringToIntegerConversion.h>
+
+namespace WebCore {
+
+static bool isInvalidIPv6HostCodePoint(auto codepoint)
+{
+    std::array<UChar32, 3> invalidSpecialCodepoints = { '[', ']', ':' };
+    return !isASCIIHexDigit(codepoint) && std::find(invalidSpecialCodepoints.begin(), invalidSpecialCodepoints.end(), codepoint) != invalidSpecialCodepoints.end();
+}
+
+// https://urlpattern.spec.whatwg.org/#is-an-absolute-pathname
+bool isAbsolutePathname(StringView input, BaseURLStringType inputType)
+{
+    if (input.isEmpty())
+        return false;
+
+    if (input.startsWith("/"_s))
+        return true;
+
+    if (inputType == BaseURLStringType::URL)
+        return false;
+
+    if (input.length() < 2)
+        return false;
+
+    if (input.startsWith("\\/"_s))
+        return true;
+
+    if (input.startsWith("{/"_s))
+        return true;
+
+    return false;
+}
+
+// https://urlpattern.spec.whatwg.org/#canonicalize-a-protocol, combined with https://urlpattern.spec.whatwg.org/#process-protocol-for-init
+ExceptionOr<String> canonicalizeProtocol(String&& value, BaseURLStringType valueType)
+{
+    if (value.isEmpty())
+        return value;
+
+    String strippedValue = value.endsWith(":"_s) ? value.substring(0, value.length()-1) : value;
+
+    if (valueType == BaseURLStringType::Pattern)
+        return strippedValue;
+
+    URL dummyURL(makeString(strippedValue, "://webkit.test"_s));
+
+    if (!dummyURL.isValid())
+        return Exception { ExceptionCode::TypeError };
+
+    return dummyURL.protocol().toString();
+}
+
+// https://urlpattern.spec.whatwg.org/#canonicalize-a-username, combined with https://urlpattern.spec.whatwg.org/#process-username-for-init
+String canonicalizeUsername(String&& value, BaseURLStringType valueType)
+{
+    if (value.isEmpty())
+        return value;
+
+    if (valueType == BaseURLStringType::Pattern)
+        return value;
+
+    URL dummyURL("https://www.webkit.org"_s);
```

Comment:
Hm... this will create a dummy URL to canonicalize each component. That seems awfully inefficient.
Not that perf matters here but perhaps these functions should form a class with a URL as an instance variable.

Reply:
On the other hand, spec does say to create dummy URL in each canonicalization step so maybe the current code is okay.